upgrade-2.x.sh

```diff
@@ -696,6 +701,61 @@ function find_partitions {
     log "Update partition: ${update_part}"
 }
 
+#######################################
+# Query public apps for a matching image
+# Globals:
+#   APIKEY
+#   API_ENDPOINT
+#   SLUG
+#   VARIANT
+# Arguments:
+#   version: the OS version to look for
+# Returns:
+#   Registry URL for desired image
+#######################################
+function get_image_location() {
+    local version=$1
+    # TODO: could improve the quality of the API call here
+    curl --retry 10 --silent -X GET \
+        -H "Content-Type: application/json" \
+        -H "Authorization: Bearer ${APIKEY}" \
+        "${API_ENDPOINT}/v5/release?\$expand=release_tag,belongs_to__application,contains__image/image&\$filter=(belongs_to__application/any(a:a/device_type%20eq%20'${SLUG}'))%20and%20(release_tag/any(rt:(rt/tag_key%20eq%20'version')%20and%20(rt/value%20eq%20'${version}')))" \
+        | jq -r ".d[] | select(.release_tag[].value == (\"${VARIANT}\" | ascii_downcase)) | .contains__image[0].image[0] | [.is_stored_at__image_location, .content_hash] | \"\(.[0])@\(.[1])\""
+}
+
+#######################################
+# Find a delta in the registry between two hostapp versions using the API
+# Globals:
+#   APIKEY
+#   DELTA_ENDPOINT
+#   DELTA_VERSION
+#   VERSION
+# Arguments:
+#   target_version: the desired OS target version
+# Returns:
+#   Location of delta image
+#######################################
+function find_delta() {
+    # we need to strip the target_version's variant tag to query the API properly
+    local target_version=${1/.dev/}
+    local src_image target_image
+    # shellcheck disable=SC2153
+    src_image=$(get_image_location "${VERSION}")
+    target_image=$(get_image_location "${target_version}")
+    if [ -z "${src_image}" ] || [ -z "${target_image}" ]; then
+        return
+    else
+        # TODO: should we retry this more extensively? deltas may take a while to generate..
+        delta=$(curl --retry 10 --silent -X GET \
+            "${DELTA_ENDPOINT}/api/v${DELTA_VERSION}/delta?src=${src_image}&dest=${target_image}" \
```

Comment:
This needs to be retrying significantly more than 10 times, deltas can take an arbitrary amount of time to be generated on demand, typically depending on the sizes of the images. You’ll also ideally backoff a bit. Lastly, for any response code other than 200 and 503, you’ll need to stop polling and either fail the script or skip deltas.

On the other hand, it’s going to be far better if deltas are pregenerated when the OS releases are made, so this call returns immediately.

Reply:
There's an open question on whether or not we want to pregenerate the deltas -- probably best to discuss in the mini-spec (https://github.com/balena-io/balena/pull/2127)?